### PR TITLE
Fix #513, Order of fields in ParamDefinition

### LINKED ISSUE #513: Order of fields in ParamDefinition
The current syntax is: 
```
  if (param_definition_mode == 0) {
    leb128() duration;
    leb128() num_subblocks;
    leb128() constant_subblock_duration;
    if (constant_subblock_duration == 0) {
      for (i=0; i< num_subblocks; i++) {
        leb128() subblock_duration;
      }
    } 
  }
```
Why is `num_subblocks` stored before `constant_subblock_duration` and not in the `if` clause?

### PR DESCRIPTION
**Pull request by sunghee-hwang**
***
<a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/iamf/pull/529.html" title="Last updated on Jun 27, 2023, 7:38 PM UTC (6a7f7bc)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/iamf/529/3318721...6a7f7bc.html" title="Last updated on Jun 27, 2023, 7:38 PM UTC (6a7f7bc)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -779,14 +779,14 @@ abstract class ParamDefinition() {
...
-
 <dfn noexport>constant_subblock_duration</dfn> specifies the duration of each subblock, in the case where all subblocks except the last subblock have equal durations. If all subblocks except the last subblock do not have equal durations, the value of constant_subblock_duration SHALL be set to 0. 
 
-When it defines <dfn noexport>D</dfn> = the value of [=duration=], <dfn noexport>NS</dfn> = the value of [=num_subblocks=], <dfn noexport>CSI</dfn> = the value of [=constant_subblock_duration=] and <dfn noexport>SI</dfn> = the value of [=subblock_duration=].
-- When [=CSI=] != 0, [=NS=] x [=CSI=] SHALL be equal to or greater than [=D=].
-	- If [=NS=] x [=CSI=] > [=D=], the actual duration of the last subblock SHALL be [=D=] - ([=NS=] - 1) x [=CSI=].
-- When [=CSI=] = 0, the summation of all [=SI=]s in this parameter block SHALL be equal to [=D=].
+When it defines <dfn noexport>D</dfn> = the value of [=duration=], <dfn noexport>NS</dfn> = the value of [=num_subblocks=], <dfn noexport>CSD</dfn> = the value of [=constant_subblock_duration=] and <dfn noexport>SD</dfn> = the value of [=subblock_duration=].
```
Comment (cconcolato) [inline index.bs:784]: I'm not sure I understand the meaning of "When it defines [...]". Do you mean "Given [...], the following constraints apply:" ?

File: index.bs [hunk 2]
```diff
@@ -779,14 +779,14 @@ abstract class ParamDefinition() {
...
+- When [=CSD=] != 0, [=NS=] = roundup([=D=] / [=CSD=]).
```
Comment (sunghee-hwang) [inline index.bs:785]:
```suggestion
- When [=CSD=] != 0, [=num_subblocks=] is implicitly calculated by the equation, [=NS=] = roundup([=D=] / [=CSD=]).
```

File: index.bs [hunk 3]
```diff
@@ -1399,10 +1403,10 @@ class parameter_block_obu() {
 
 <dfn value noexport for="parameter_block_obu()">duration</dfn> specifies the duration for which this parameter block is valid and applicable. 
 
-<dfn value noexport for="parameter_block_obu()">num_subblocks</dfn> specifies the number of different sets of parameter values specified in this parameter block, where each set describes a different subblock of the timeline, contiguously.
-
 <dfn value noexport for="parameter_block_obu()">constant_subblock_duration</dfn> specifies the duration of each subblock, in the case where all subblocks except the last subblock have equal durations. If all subblocks except the last subblock do not have equal durations, the value of constant_subblock_duration SHALL be set to 0. 
 
+<dfn value noexport for="parameter_block_obu()">num_subblocks</dfn> specifies the number of different sets of parameter values specified in this parameter block, where each set describes a different subblock of the timeline, contiguously.
```
Comment (sunghee-hwang) [inline index.bs:1408]:
```suggestion
<dfn value noexport for="parameter_block_obu()">num_subblocks</dfn> specifies the number of different sets of parameter values specified in this parameter block, where each set describes a different subblock of the timeline, contiguously. When [=constant_subblock_duration=] != 0, [=num_subblocks=] is implicitly calculated by the equation,   [=num_subblocks=] = roundup([=duration=]/[=constant_subblock_duration=]).
```

**Review by sunghee-hwang — COMMENTED** [review]
More clarification

**Review by cconcolato — APPROVED** [review]

File: index.bs [hunk 4]
```diff
@@ -779,14 +779,14 @@ abstract class ParamDefinition() {
...
 
-When it defines <dfn noexport>D</dfn> = the value of [=duration=], <dfn noexport>NS</dfn> = the value of [=num_subblocks=], <dfn noexport>CSI</dfn> = the value of [=constant_subblock_duration=] and <dfn noexport>SI</dfn> = the value of [=subblock_duration=].
-- When [=CSI=] != 0, [=NS=] x [=CSI=] SHALL be equal to or greater than [=D=].
-	- If [=NS=] x [=CSI=] > [=D=], the actual duration of the last subblock SHALL be [=D=] - ([=NS=] - 1) x [=CSI=].
-- When [=CSI=] = 0, the summation of all [=SI=]s in this parameter block SHALL be equal to [=D=].
+Given <dfn noexport>D</dfn> = the value of [=duration=], <dfn noexport>NS</dfn> = the value of [=num_subblocks=], <dfn noexport>CSD</dfn> = the value of [=constant_subblock_duration=] and <dfn noexport>SD</dfn> = the value of [=subblock_duration=].
+- When [=CSD=] != 0, [=num_subblocks=] is implicitly calculated by the equation, [=NS=] = roundup([=D=] / [=CSD=]).
```
Comment (felicialim) [inline index.bs:785]:
```suggestion
- When [=CSD=] != 0, [=num_subblocks=] is implicitly calculated as [=NS=] = roundup([=D=] ÷ [=CSD=]).
```

Sec 8.2. Arithmetic Operators defines `/` as "Integer division with truncation of the result toward zero"

File: index.bs [hunk 5]
```diff
@@ -1367,8 +1367,12 @@ class parameter_block_obu() {
   
   if (param_definition_mode) {
     leb128() duration;
-    leb128() num_subblocks;
     leb128() constant_subblock_duration;
+    if (constant_subblock_duration == 0) {
+      leb128() num_subblocks;
+    } else {
+      // num_subblocks = roundup(duration / constant_subblock_duration)
```
Comment (felicialim) [inline index.bs:1374]:
```suggestion
      // num_subblocks = roundup(duration ÷ constant_subblock_duration)
```

File: index.bs [hunk 6]
```diff
@@ -1399,10 +1403,10 @@ class parameter_block_obu() {
 
 <dfn value noexport for="parameter_block_obu()">duration</dfn> specifies the duration for which this parameter block is valid and applicable. 
 
-<dfn value noexport for="parameter_block_obu()">num_subblocks</dfn> specifies the number of different sets of parameter values specified in this parameter block, where each set describes a different subblock of the timeline, contiguously.
-
 <dfn value noexport for="parameter_block_obu()">constant_subblock_duration</dfn> specifies the duration of each subblock, in the case where all subblocks except the last subblock have equal durations. If all subblocks except the last subblock do not have equal durations, the value of constant_subblock_duration SHALL be set to 0. 
 
+<dfn value noexport for="parameter_block_obu()">num_subblocks</dfn> specifies the number of different sets of parameter values specified in this parameter block, where each set describes a different subblock of the timeline, contiguously. When [=constant_subblock_duration=] != 0, [=num_subblocks=] is implicitly calculated by the equation,   [=num_subblocks=] = roundup([=duration=]/[=constant_subblock_duration=]).
```
Comment (felicialim) [inline index.bs:1408]:
```suggestion
<dfn value noexport for="parameter_block_obu()">num_subblocks</dfn> specifies the number of different sets of parameter values specified in this parameter block, where each set describes a different subblock of the timeline, contiguously. When [=constant_subblock_duration=] != 0, [=num_subblocks=] is implicitly calculated as  [=num_subblocks=] = roundup([=duration=] ÷ [=constant_subblock_duration=]).
```

**Review by felicialim — APPROVED** [review]

**Review by tdaede — APPROVED** [review]

**Pull request merged**